source/generic.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// genericSource is an implementation of Source that provides endpoints by listing
+// specified CRD and fetching Endpoints embedded in Spec.
+type genericSource struct {
+	crdClient rest.Interface
+	namespace string
+	crdKind   string
+	scheme    *runtime.Scheme
+}
+
+func addKnownTypes(scheme *runtime.Scheme, groupVersion schema.GroupVersion) error {
+	scheme.AddKnownTypes(groupVersion,
+		&endpoint.DNSEndpoint{},
+		&endpoint.DNSEndpointList{},
+	)
+	metav1.AddToGroupVersion(scheme, groupVersion)
+	return nil
+}
+
+// NewCRDClientForAPIVersionKind return rest client for the given apiVersion and kind of the CRD
+func NewCRDClientForAPIVersionKind(client kubernetes.Interface, kubeConfig, kubeMaster, apiVersion, kind string) (*rest.RESTClient, *runtime.Scheme, error) {
+	if kubeConfig == "" {
```

Comment:
Is this the same as using the in-cluster config?

Reply:
yes, it is almost the same. if user has not set the kubeconfig path explicitly, it would look into standard path `$HOME/.kube/config` and if that is also not present then uses in-cluster config. I just reused the logic already done by external-dns in here https://github.com/kubernetes-incubator/external-dns/blob/master/source/store.go#L111